### ISSUE TITLE
Can not map classes with properties of type EnumSet

### ISSUE THREAD

**Issue by ikettu** [modelmapper/modelmapper#260]
[EnumSetMappingTest.zip](https://github.com/modelmapper/modelmapper/files/1155640/EnumSetMappingTest.zip)
Trying to map classes with similar EnumSets fails with exception. Attached test works on ModelMapper version 0.7.6 but not 1.1.0

I have tried modify converter and tried to configure provider but have not had any success.

> org.modelmapper.ConfigurationException: ModelMapper configuration errors:
> 
> 1) Failed to generate proxy class for java.util.EnumSet. Ensure that java.util.EnumSet has a non-private constructor.
> Caused by: java.lang.IllegalArgumentException: No visible constructors in class java.util.EnumSet
> 	at org.modelmapper.internal.cglib.proxy.Enhancer.filterConstructors(Enhancer.java:666)
> 	at org.modelmapper.internal.cglib.proxy.Enhancer.generateClass(Enhancer.java:567)
> 	at org.modelmapper.internal.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
> 	at org.modelmapper.internal.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:329)
> 	at org.modelmapper.internal.cglib.proxy.Enhancer.generate(Enhancer.java:492)
>         ...

Test:
```java
package fi.fastroi.merp.util.mapper;

import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.assertThat;

import java.util.EnumSet;

import org.junit.BeforeClass;
import org.junit.Test;
import org.modelmapper.AbstractProvider;
import org.modelmapper.Converter;
import org.modelmapper.ModelMapper;
import org.modelmapper.PropertyMap;
import org.modelmapper.Provider;
import org.modelmapper.spi.MappingContext;


public class EnumSetMappingTest {

	public static enum E {
		E1, E2, E3, E4
	}
	
	public static class A {
		private EnumSet<E> enumsA;

		public EnumSet<E> getEnumsA() {
			return enumsA;
		}
		public void setEnumsA(EnumSet<E> enumsA) {
			this.enumsA = enumsA;
		} 
	}
	

	public static class B {
		private EnumSet<E> enumsB;

		public EnumSet<E> getEnumsB() {
			return enumsB;
		}
		public void setEnumsB(EnumSet<E> enumsB) {
			this.enumsB = enumsB;
		} 
	}


	
	private static ModelMapper mapper = new ModelMapper();
	
	@BeforeClass
	public static void initMapper() {
		mapper.addMappings(new PropertyMap<A, B>() {
			
			Converter<EnumSet<?>, EnumSet<?>> enumSetConverter = new Converter<EnumSet<?>, EnumSet<?>>() {
				@Override
				public EnumSet<?> convert(MappingContext<EnumSet<?>, EnumSet<?>> context) {
				    Object source = context.getSource();
				    if (source == null)
				      return null;

				    return EnumSet.copyOf((EnumSet<?>) source);
				}

			};
			
//			Provider<EnumSet<E>> enumSetProvider = new AbstractProvider<EnumSet<E>>() {
//
//				@Override
//				protected EnumSet<E> get() {
//					return EnumSet.noneOf(E.class);
//				}
//			};

			@Override
			protected void configure() {
				using(enumSetConverter).map(source.getEnumsA()).setEnumsB(null);
				
//				 with(enumSetProvider).using(enumSetConverter).map().setEnumsB(source.getEnumsA());
			}
		});
		
	}

	
	@Test
	public void testEnumSetWithValues() {
	
		A a = new A();
		a.enumsA = EnumSet.of(E.E1, E.E2);
		B b = mapper.map(a, B.class);
		mapper.validate();

		assertThat(b.getEnumsB(), hasItems(E.E1, E.E2));

	}
		
		
	@Test
	public void testEnumSetAsEmpty() {
		
		A a = new A();
		B b = mapper.map(a, B.class);
		mapper.validate();

		assertThat(b.getEnumsB(), is(nullValue()));
	}
}

```

**Comment by chhsiao90** [comment]
It looks like 5933cf1215d5ac35cdce6ec83edf3787140ebde7 break the test after I used git bisect to determine the bad commit. I will take a look what's happen.

**Comment by chhsiao90** [comment]
@ikettu - I think it's some restriction with the usage of cglib please check the commit https://github.com/cglib/cglib/commit/c070390c71810f3137b4ef27e1827f2784f1ee2f.

We used some proxy methodology for generating explicit mapping, and we used cglib for our methodology. And cglib think that some members that match 1. package-private 2. in BootstrapClassloader should not be proxied. So with the restriction, we cannot map EnumSet anymore.

We don't plan to downgrade cglib, and I don't think it's a bug for cglib.
But there is still a workaround to fix your problem, you can use Set<E> instead of EnumSet<E> when you declared your class. And the result works fine.

```java
	public static class A {
		private Set<E> enumsA;

		public Set<E> getEnumsA() {
			return enumsA;
		}
		public void setEnumsA(Set<E> enumsA) {
			this.enumsA = enumsA;
		} 
	}

	public static class B {
		private Set<E> enumsB;

		public Set<E> getEnumsB() {
			return enumsB;
		}
		public void setEnumsB(Set<E> enumsB) {
			this.enumsB = enumsB;
		} 
	}
```

**Comment by ikettu** [comment]
Using Set instead of EnumSet is not really workaround here when you have existing entities and DTO's already. Any way to get around this by using just providers/converters ??

But in my opinion, since EnumSet is part of Java, mapper should handle it correctly.

**Comment by chhsiao90** [comment]
Got it, that make sense, I will try to find out a better solution.

**Comment by chhsiao90** [comment]
Fixed by #266 and will be released in next version.

**Comment by chhsiao90** [comment]
The bugfix had been released with ModelMapper v1.1.1, please take a try, thanks!
